1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+    then `TypeResolver` should use the cache associated with ALC-M.
+
+  - Another example, if a `ITypeName` has already resolved the type via reflection in default-ALC,
+    and the `ScriptBlock` containing that `ITypeName` is now being executed in the context of an isolated module M,
+    then the cached type referred by `ITypeName` should be invalided and another resolution should be triggered.
+    Similarly, if the `ScriptBlock` has already been compiled into LINQ expression tree and have the delegate generated,
+    then the cached delegate needs to be voided and another compilation needs to be enforced to make sure type resolution is correct.
+
+For the latter category, theoretically they will continue to work without change,
+because types loaded into different load contexts are different.
+
+### Reclaiming Module AssemblyLoadContext
+
+Starting from 3.0, .NET Core supports unloading an `AssemblyLoadContext` and all assemblies in it.
+Ideally, we would like to reclaim the load context associated with an isolated module when the module is removed.
+
+The cache updates called out in the [Update Type Resolution Logic](#update-type-resolution-logic) section
+were designed with the purpose of allowing the custom ALC to be reclaimed.
+By having each custom ALC to hold the cache data structure within itself,
+we won't need to hold a custom ALC or any types loaded in a custom ALC as the key or value for any caches in our engine code,
+and thus there is no need to invalidate any engine cache.
+
+However, we won't be able to guarantee that a custom ALC could be unloaded at all.
+This is because Dynamic Language Runtime (DLR) holds types in its cache, indefinitely.
+
+The DLR types live in the default-ALC,
+and it will generate a `DynamicMetaObject` object for a late-binding operation,
+which is something like a key-value pair `<restrictions, expression>` that will be held in the DLR caches.
+This is a perf improvement feature for dynamic languages like PowerShell because it allows you to skip an expensive late-binding operation when the `restrictions` matches.
+That key-value pair will be held indefinitely,
+so once types from a custom load context is involved in a late binding operation,
+such as property access, method invocation and etc,
+it will be rooted in the GC Heap by the DLR cache, and the custom load context cannot be unloaded afterwards.
+
+## Summary
+
+For the initial implementation, here are the goals and non-goals:
+
+- Goals: work items called out in the [Specification](#specification) section,
+  including updates to the module cmdlets, assembly loading, and type resolution.
+- Non-goals: the new type reference syntax.
+
+To emphasize again:
+The type identity issue is inevitable as soon as different versions of the same assembly can be loaded in the same PowerShell session.
+This will mainly affect modules that expect users to directly create objects of the types loaded by the modules.
```

Comment:
> It seems we forget about exception processing coming from different ALC-s.

@iSazonov Can you be more specific what exception processing is needed?

Reply:
It comes from https://docs.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions#ensure-that-exception-data-is-available-when-code-executes-remotely

If there is a cascading exception processing it can be not trivial how we set/restore correct ALC and at the end I wonder how Get-Error will works (I say about exposing to user in common.).

Reply:
This is interesting. I will experiment and check it out if the exception would be a problem.

Reply:
I have verified and the across-ALC exception handling is not a problem. I think that makes sense, as they are all in the same AppDomain.

**simple.dll**
```c#
using System;

namespace simple
{
    public class Class1 : ICloneable
    {
        public static void Run(bool throwException)
        {
            if (throwException)
            {
                throw new MyCustomException();
            }

            Console.WriteLine("Run !!!");
        }

        public object Clone()
        {
            Console.WriteLine("~~~ Clone is about to throw ~~~");
            throw new MyCustomException();
        }
    }

    public class MyCustomException : Exception
    {
        public string Secret { get; }

        internal MyCustomException()
        {
            Secret = "This is a secret";
        }
    }
}
```

**Load simple.dll into separate ALC and execute method that throws the custom exception**
```c#
using System;
using System.Reflection;
using System.Runtime.Loader;

namespace alc_2
{
    class Program
    {
        static void Main(string[] args)
        {
            MyALC myALC = new();

            Assembly asm = myALC.LoadFromAssemblyPath(@"C:\arena\apps\alc-2\simple.dll");
            Type type = asm.GetType("simple.Class1");

            var cloneable = (ICloneable) Activator.CreateInstance(type);

            try
            {
                cloneable.Clone();
            }
            catch (TargetInvocationException e)
            {
                Console.WriteLine("-- TargetInvocationException");
                Console.WriteLine($"-- InnerException: {e.InnerException.GetType().FullName}");
            }
            catch (Exception e)
            {
                Console.WriteLine($"-- exception: {e.GetType().FullName}");
            }

            Console.WriteLine("---- done in master");
        }
    }

    internal class MyALC : AssemblyLoadContext
    {
        protected override Assembly Load(AssemblyName assemblyName)
        {
            Console.WriteLine("==== MyALC-Load ====");
            Console.WriteLine($"   |--- {assemblyName.FullName}");
            return null;
        }
    }
}
```
**result**
```
PS:97> dotnet run
==== MyALC-Load ====
   |--- System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
==== MyALC-Load ====
   |--- System.Console, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
~~~ Clone is about to throw ~~~
-- exception: simple.MyCustomException
---- done in master
```

Reply:
You do all work in **one** assembly.
From docs:
> If App Domain B throws an exception that is contained in an assembly under its application base, but not under App Domain A's application base, App Domain A will not be able to find the exception, and the common language runtime will throw a FileNotFoundException exception.

Reply:
Not sure what you mean specifically.
- There is only 1 app domain in .NET Core. 
- 2 assemblies were involved in my check, `simple.dll` and the executable that contains the main method.
- `simple.dll` is in a path that cannot be found by the default ALC.

Maybe you want to try it yourself?

Reply:
Perhaps I think about too complex scenario no needed for the RFC purposes.

Load NewtonSoftV1 in ALC1, NewtonSoftV2 in ALC2 , throw in ALC1 - sometime there could be weird cases (we are in ALC2 context or in Default context) when we can not catch the exception since there is really two exceptions.

Reply:
I presume you mean the scenario where the same exception type exists in both the default ALC and the custom ACL (from 2 versions of the same assembly). When executing code from the assembly loaded in the custom ALC, the exception gets thrown, and you wonder whether we are able to catch that exception correctly from the code loaded in the default ALC.

It makes sense to check on this scenario, and the answer is yes, the exception can be caught correctly. See the code and results below:

<details>
  <summary>Click to expand the code snippet</summary>
  
  ```c#
using System;
using System.Reflection;
using System.Runtime.Loader;

namespace alc_2
{
    class Program
    {
        static void Main(string[] args)
        {
            MyALC myALC = new();

            Assembly asm10 = myALC.LoadFromAssemblyPath(@"C:\arena\apps\alc-2\v1.0\simple.dll");
            Type type1 = asm10.GetType("simple.Class1");

            var cloneable1 = (ICloneable) Activator.CreateInstance(type1);
            Console.WriteLine("Run code from simple.dll v1.0 loaded in a custom ALC");
            Run(cloneable1);

            Console.WriteLine("\n");

            Assembly asm12 = AssemblyLoadContext.Default.LoadFromAssemblyPath(@"C:\arena\apps\alc-2\v1.2\simple.dll");
            Type type2 = asm12.GetType("simple.Class1");

            var cloneable2 = (ICloneable) Activator.CreateInstance(type2);
            Console.WriteLine("Run code from simple.dll v1.2 loaded in the default ALC");
            Run(cloneable2);

            Console.WriteLine("---- done in master");
        }

        public static void Run(ICloneable cloneable)
        {
            try
            {
                cloneable.Clone();
            }
            catch (TargetInvocationException e)
            {
                Console.WriteLine("--- TargetInvocationException ---");
                Console.WriteLine($"InnerException: {e.InnerException.GetType().FullName}");
                Console.WriteLine(e.InnerException.ToString());
                Console.WriteLine("------------- End ---------------");
            }
            catch (Exception e)
            {
                Console.WriteLine("---------- Exception -----------");
                Console.WriteLine(e.GetType().FullName);
                Console.WriteLine(e.ToString());
                Console.WriteLine("------------ End ---------------");
            }
        }
    }

    internal class MyALC : AssemblyLoadContext
    {
        protected override Assembly Load(AssemblyName assemblyName)
        {
            Console.WriteLine("==== MyALC-Load ====");
            Console.WriteLine($"   |--- {assemblyName.FullName}");
            return null;
        }
    }
}
  ```
</details>

**Results:**
```
PS> dotnet run
==== MyALC-Load ====
   |--- System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Run code from simple.dll v1.0 loaded in a custom ALC
==== MyALC-Load ====
   |--- System.Console, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
~~~ Clone is about to throw ~~~
---------- Exception -----------
simple.MyCustomException
simple.MyCustomException from the simple.dll (1.0.0.0) loaded in custom ALC
------------ End ---------------


Run code from simple.dll v1.2 loaded in the default ALC
~~~ Clone is about to throw ~~~
---------- Exception -----------
simple.MyCustomException
simple.MyCustomException from the simple.dll (1.2.0.0) loaded in default ALC
------------ End ---------------
---- done in master
```

Reply:
I feel the potential benefit of this proposal is worth continued investigation.  Admittedly it won't solve all version conflict problems, but I think we are at the point where prototyping is needed.  I agree with initially keeping the feature as simple as possible.  It will be interesting to see how well it solves versioning problems and what, if any, breaking/blocking problems arise.

Reply:
I do not wonder C# code can catch an exception since Exception class is defined in Runtime and shared from default ALC.
But what about:
```powershell
try {
    cmdletFromDefaultALC | cmdletFromALC1
} catch [simple::MyCustomException] {
    // my concern about explicit type - user can be amazed why he can not catch the exception.
    // and what do Get-Error show at end?
}

Reply:
If `MyCustomException` exists in both default ALC and ALC-1, then `[simple::MyCustomException]` in this case is resolved to the one from default ALC, and thus it can only catches if such an exception is thrown from `cmdletFromDefaultALC`.

If `cmdletFromALC1` throws `MyCustomException`, then it won't be caught. `Get-Error` will show the `MyCustomException` though, because it's still catchable in C#.